### PR TITLE
[circleci] Enable ewasm tests.

### PR DESCRIPTION
**Pull request by aarlt**
Replaces #9212.

### CODE REVIEW

**Comment by aarlt** [comment]
For now: run soltest ewasm tests without optimization only. See https://gist.github.com/aarlt/ff0ff4d453bd0c1f708e82c0ac82cad6#file-soltest-ewasm-error

File: .circleci/soltest.sh [hunk 1]
```diff
@@ -47,15 +47,28 @@ mkdir -p test_results
...
+  LIBHERA=/usr/local/lib/libhera.dylib
+else
+  LIBEVMONE=/usr/lib/libevmone.so
+  LIBHERA=/usr/lib/libhera.so
+fi
+
+BOOST_TEST_ARGS="--color_output=no --show_progress=yes --logger=JUNIT,error,test_results/$(get_logfile_basename).xml ${BOOST_TEST_ARGS}"
+SOLTEST_ARGS="--evm-version=$EVM $SOLTEST_FLAGS --vm ${LIBEVMONE}"
+
+# run tests against hera ewasm evmc vm byzantium only if OPTIMIZE == 0
```
Comment (chriseth) [inline .circleci/soltest.sh]: Can you move this to soltest_all.sh instead?

File: .circleci/soltest.sh [hunk 2]
```diff
@@ -47,15 +47,28 @@ mkdir -p test_results
...
+  test "${OPTIMIZE}" = "1" && filename="${filename}_opt"
+  test "${ABI_ENCODER_V2}" = "1" && filename="${filename}_abiv2"
 
-    echo -ne "${filename}"
+  echo -ne "${filename}"
 }
 
-BOOST_TEST_ARGS="--color_output=no --show_progress=yes --logger=JUNIT,error,test_results/`get_logfile_basename`.xml ${BOOST_TEST_ARGS}"
-SOLTEST_ARGS="--evm-version=$EVM $SOLTEST_FLAGS"
+if [[ "$OSTYPE" == "darwin"* ]]; then
```
Comment (leonardoalt) [inline .circleci/soltest.sh]: Were the semantic tests not run on osx before?

File: .circleci/soltest.sh (hunk 1 above)
Comment (aarlt) [reply .circleci/soltest.sh]: Yes, sure.

File: .circleci/soltest.sh (hunk 2 above)
Comment (aarlt) [reply .circleci/soltest.sh]: Yes, the semantic tests where also executed on osx before.

In the past the finding of the evmone shared library was done implicitly. But this implicit searching for emvone is disabled, if a `--vm` argument is supplied. The shared libraries need then to be defined explicitly. The main reason for this is that we decided not only to support evmone or hera, the `--vm` parameter can be used to define any evmc compatible shared library.

File: .circleci/config.yml [hunk 3]
```diff
@@ -588,6 +588,25 @@ jobs:
       - store_test_results: *store_test_results
       - store_artifacts: *artifacts_test_results
 
+  test_osx:
```
Comment (chriseth) [inline .circleci/config.yml]: I'm not sure we really need this. If the bytecode comparison works, then this is actually redundant.

**Review by chriseth — CHANGES_REQUESTED** [review]
Can you also take a look at the other (non-circleci) scripts that execute the tests like `scripts/tests.sh`?

**Comment by leonardoalt** [comment]
Ping: reviewer requested changes

File: .circleci/config.yml (hunk 3 above)
Comment (aarlt) [reply .circleci/config.yml]: ok, removed.

**Comment by aarlt** [comment]
> Can you also take a look at the other (non-circleci) scripts that execute the tests like `scripts/tests.sh`?

@chriseth I updated `scripts/tests.sh`, there seem to be no other other related scripts.

File: .circleci/soltest_all.sh [hunk 4]
```diff
@@ -70,6 +70,12 @@ do
     for EVM in ${EVM_VALUES[@]}
     do
         [[ " $RUN_STEPS " =~ " $STEP " ]] && EVM="$EVM" OPTIMIZE="$OPTIMIZE" BOOST_TEST_ARGS="-t !@nooptions" "${REPODIR}/.circleci/soltest.sh"
+        EWASM_ARGS=""
+
+        # run tests against hera ewasm evmc vm, only if OPTIMIZE == 0 and evm version is byzantium
```
Comment (chriseth) [inline .circleci/soltest_all.sh:72]: Is the idea that we only run a single combination on ewasm? Then I would recommand to not use `soltest_all.sh` but `soltest.sh` directly.

File: scripts/tests.sh [hunk 5]
```diff
@@ -115,8 +115,11 @@ do
                 fi
             fi
 
+            EWASM_ARGS=""
+            [ "${vm}" = "byzantium" ] && [ "${optimize}" = "" ] && EWASM_ARGS="--ewasm"
+
             set +e
-            "${SOLIDITY_BUILD_DIR}"/test/soltest --show-progress $log -- --testpath "$REPO_ROOT"/test "$optimize" --evm-version "$vm" $SMT_FLAGS $force_abiv2_flag
+            "${SOLIDITY_BUILD_DIR}"/test/soltest --show-progress $log -- ${EWASM_ARGS} --testpath "$REPO_ROOT"/test "$optimize" --evm-version "$vm" $SMT_FLAGS $force_abiv2_flag
```
Comment (chriseth) [inline scripts/tests.sh:122]: Why not run them after the loop?

File: .circleci/soltest_all.sh [hunk 6]
```diff
@@ -69,7 +69,11 @@ for OPTIMIZE in ${OPTIMIZE_VALUES[@]}
 do
     for EVM in ${EVM_VALUES[@]}
     do
-        [[ " $RUN_STEPS " =~ " $STEP " ]] && EVM="$EVM" OPTIMIZE="$OPTIMIZE" BOOST_TEST_ARGS="-t !@nooptions" "${REPODIR}/.circleci/soltest.sh"
+        # run tests against hera ewasm evmc vm, only if OPTIMIZE == 0 and evm version is byzantium
+        EWASM_ARGS=""
+        [ "${EVM}" = "byzantium" ] && [ "${OPTIMIZE}" = "0" ] && EWASM_ARGS="--ewasm"
+
+        [[ " $RUN_STEPS " =~ " $STEP " ]] && EVM="$EVM" OPTIMIZE="$OPTIMIZE" SOLTEST_FLAGS="$EWASM_ARGS" BOOST_TEST_ARGS="-t !@nooptions" "${REPODIR}/.circleci/soltest.sh"
```
Comment (chriseth) [inline .circleci/soltest_all.sh]: This disables any SOLTEST_FLAGS passed in from outside, right? Shouldn't we change it to `SOLTEST_FLAGS="$SOLTEST_FLAGS $EWASM_ARGS"`?

Comment (aarlt) [reply .circleci/soltest_all.sh]: Thats true

**Review by chriseth — APPROVED** [review]

**Pull request merged**